### PR TITLE
[schedulers]: get stores only once 

### LINKED ISSUE #3744: Performance issues of balance-region-scheduler
## Development Task

[tests](https://gist.github.com/disksing/a1bf78007b674da7e15b52a9e25eb3a5) show that when there are too many stores in a cluster, `balance-region-scheduler` could have performance issues.

- [ ] Improve ruleChecker (waiting regions too long)
- [ ] Improve fitRegion
- [x] Improve `GetStoreRegionCount`
- [ ] Improve the big loop of `BalanceRegionScheduler`

### PR DESCRIPTION
**Pull request by bufferflies**
Signed-off-by: bufferflies <1045931706@qq.com>

### What problem does this PR solve?
fix #3744,  Get stores only once to reduce scheduler execute time 
image

### What is changed and how it works?
stores should get once before schedule  and it should input to plan 
### Check List

Tests

- Unit test
- Manual test (add detailed scripts or steps below)

Code changes

Side effects

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tidb-ansible`](https://github.com/pingcap/tidb-ansible):
- Need to cherry-pick to the release branch

### Release note

```release-note
Please add a release note.

If you don't think this PR needs a release note then fill it with None.
```

### CODE REVIEW

**Review by lhy1024 — APPROVED** [review]

**Review by disksing — APPROVED** [review]

File: server/core/store.go [hunk 1]
```diff
@@ -618,6 +630,19 @@ func (s *StoresInfo) AttachAvailableFunc(storeID uint64, limitType storelimit.Ty
 	}
 }
 
+func (s *StoresInfo) GetStoreSet() *StoreSet {
+	stores := make([]*StoreInfo, 0, len(s.stores))
+	items := make(map[uint64]*StoreInfo, len(s.stores))
+	for id, store := range s.stores {
+		stores = append(stores, store)
```
Comment (bufferflies) [inline server/core/store.go:637]: should it need to clone

File: server/schedule/opt/opts.go [hunk 2]
```diff
@@ -47,6 +47,25 @@ type Cluster interface {
...
+	stores []*core.StoreInfo
+}
+
+// GetStores returns store infos
+func (c *CacheCluster) GetStores() []*core.StoreInfo {
+	return c.stores
+}
+
+// NewCacheCluster constructor for cache
+func NewCacheCluster(c Cluster) *CacheCluster {
```
Comment (HunDunDM) [inline server/schedule/opt/opts.go:62]:
```suggestion
func NewCacheCluster(c Cluster) Cluster {
```

File: server/schedule/opt/opts.go [hunk 3]
```diff
@@ -47,6 +47,25 @@ type Cluster interface {
 	AddSuspectRegions(ids ...uint64)
 }
 
+// CacheCluster include some cluster info
+type CacheCluster struct {
```
Comment (HunDunDM) [inline server/schedule/opt/opts.go:51]:
```suggestion
type cacheCluster struct {
```

**Review by HunDunDM — COMMENTED** [review]
rest LGTM

**Comment by HunDunDM** [comment]
Please fix the test

**Review by HunDunDM — APPROVED** [review]

**Review by nolouch — APPROVED** [review]

**Pull request merged**